### ISSUE TITLE
Paramiko transport issue

### ISSUE THREAD

**Issue by kinow** [BSC-ES/autosubmit#1323]
In GitLab by @abaer-kit on Jun 11, 2024, 15:39

Hello @dbeltrankyl and @kinow,

### Autosubmit Version

4.1.7

### Summary

I get connection errors on *some platforms* (in my case HoreKa at KIT) with the new 4.1.7 version:
```
ERROR:paramiko.transport:Secsh channel 2 open FAILED: open failed: Connect failed
```

### Steps to reproduce

Run any job on platform HoreKa.

### What is the current bug behavior?

If I try to connect to the platform twice (e.g. HOREKA, a slurm platform and HOREKA_LOGIN, a ps platform), the error already comes up when testing the connection. If I use the platform only once, it comes when trying to send the job script.

### What is the expected correct behavior?

Everything running as expected.

### Relevant logs and/or screenshots

[as-error.og](https://earth.bsc.es/gitlab/es/autosubmit/uploads/16d88a03f0ab1365c97ca83f15fc3839/as-error.og)\
[as-error-log.og](https://earth.bsc.es/gitlab/es/autosubmit/uploads/113ab8e4e91b4daf71fbe20c206a6946/as-error-log.og)\
There are a few different scenarios how this error pops up, all resulting in different traces but the same paramiko error. I can provide further details if helpful.

### Any other relevant information

Could the issue could be related to the sshd setting `MaxSessions`, which by default is 10 and on Horeka set to 2? Was there something related to this changed in 4.1.7?

Best,\
Andreas

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Jun 12, 2024, 09:05

Hello @abaer-kit , 

This part has a bug; Autosubmit only reaches this if the send command fails, fixed on 4.1.8-fixes. However this fix is not enough for your case
```
Traceback (most recent call last):
  File "/pool/autosubmit-venv/autosubmit/autosubmit/platforms/paramiko_platform.py", line 1041, in send_command
    stdin, stdout, stderr = self.exec_command(command, x11=x11)
  File "/pool/autosubmit-venv/autosubmit/autosubmit/platforms/paramiko_platform.py", line 996, in exec_command
    timeout = timeout + 60
TypeError: unsupported operand type(s) for +: 'NoneType' and 'int'
```

> Could the issue be related to the sshd setting `MaxSessions,` which by default is 10, and on Horeka set to 2? Was there something related to this change in 4.1.7?

In v4.1.4+ we changed the log retrieval mechanism to make two connections per platform ( one for sending jobs and another for retrieving the logs). But, before this rework, each log created a new connection, so I'm not sure why it happens only now.


Could you send me the experiment configuration? Does it only happen on that platform? Was this happening in 4.1.4-4.1.6?

Can you try to disable the remote logs? %PLATFORMS.PLATFORM_NAME.DISABLE_RECOVERY_THREADS%

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Jun 12, 2024, 09:08

Ah, I see that in the logs. 

`File /pool/experiments/a00d/proj/local_project/templates/art/prepare_experiment.sh` Does not exist, can you check it?

**Comment by kinow** [comment]
In GitLab by @abaer-kit on Jun 12, 2024, 09:21

> Was this happening in 4.1.4-4.1.6?

Interestingly, no. I was able to run successfully jobs with 4.1.6 (and two fixes for python executables and the stat files).

> Does it only happen on that platform?

Yes, only there.

> Can you try to disable the remote logs?

I did and still get the error.

**Comment by kinow** [comment]
In GitLab by @abaer-kit on Jun 12, 2024, 09:22

Hmmm, the file exists, I just checked.

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Jun 12, 2024, 09:27

Weird then

Can you check with a pip freeze if the autosubmitconfigparser==1.0.66? 

Can you send me the conf? I will try to reproduce it

**Comment by kinow** [comment]
In GitLab by @abaer-kit on Jun 12, 2024, 09:28

> Can you check with a pip freeze if the autosubmitconfigparser==1.0.66?

Yes, it is.

> Can you send me the conf?

Working on it :)

**Comment by kinow** [comment]
In GitLab by @abaer-kit on Jun 12, 2024, 09:33

For the configuration:\
I use a `minimal.yml`:
```yaml
CONFIG:
  # Current version of Autosubmit.
  AUTOSUBMIT_VERSION: "4.1.7"
  # Total number of jobs in the workflow.
  TOTALJOBS: 1
  # Maximum number of jobs permitted in the waiting status.
  MAXWAITINGJOBS: 1
DEFAULT:
  # Job experiment ID.
  EXPID: "a00d"
  # Default HPC platform name.
  HPCARCH: "HOREKA"
  #hint: use %PROJDIR% to point to the project folder (where the project is cloned)
  # Custom configuration location.
  CUSTOM_CONFIG: "%PROJDIR%/conf/common-KIT/art.yml"
project:
  # Type of the project.
  PROJECT_TYPE: local
  # Folder to hold the project sources.
  PROJECT_DESTINATION: local_project
local:
  PROJECT_PATH: "/home/glrunner/builds/xwfusa3BP/0/auto-icon/auto-icon"
```
and the `art.yml`:
```yaml
ICON_CASE: art
EXPNAME: NWP_EXT_DATA
DEFAULT:
  CUSTOM_CONFIG:
    PRE:
      - "%PROJDIR%/conf/common-KIT/platforms.yml"
      - "%PROJDIR%/conf/%ICON_CASE%/experiments/%EXPNAME%.yml"
```
`platforms.yml`:
```yaml
PLATFORMS:
  HOREKA:
    TYPE: slurm
    HOST: horeka.scc.kit.edu
    USER: <username>
    PROJECT: <projectname>
    SCRATCH_DIR: /home
    PARTITION: 'dev_cpuonly'
    PROCESSORS_PER_NODE: 76
    MAX_PROCESSORS: 152
    MAX_WALLCLOCK: '01:00'
    DISABLE_RECOVERY_THREADS: True
```
`NWP_EXT_DATA.yml`
```yaml
JOBS:
  PREPARE_EXPERIMENT:
    FILE: "templates/%ICON_CASE%/prepare_experiment.sh,templates/utils/bash_utils.sh"
    RUNNING: once
    WALLCLOCK: 00:05

EXPERIMENT:
  DATELIST: 20040101
  MEMBERS: "default"
  CHUNKSIZEUNIT: hour
  CHUNKSIZE: 1
  NUMCHUNKS: 1
  CHUNKINI: ''
  CALENDAR: standard
```

This should be the slightly adapted "minimal working" config. The full config is here (should be publicly accessible):\
https://gitlab.dkrz.de/auto-icon/auto-icon/-/tree/dev/as-4.1-compatibility?ref_type=heads

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Jun 12, 2024, 09:37

Thanks! I'll do some testing and report back

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Jun 12, 2024, 10:11

Can you test with 4.1.8-fixes branch?

**Comment by kinow** [comment]
In GitLab by @abaer-kit on Jun 12, 2024, 10:33

I tried, but the only difference is that it actually fails to open more channels:
```
ERROR:paramiko.transport:Secsh channel 2 open FAILED: open failed: Connect failed
ERROR:paramiko.transport:Secsh channel 3 open FAILED: open failed: Connect failed
ERROR:paramiko.transport:Secsh channel 4 open FAILED: open failed: Connect failed
```

Btw., is there a possibility to check back, that the latest "version" is actually used (as autosubmit --version will report only the official version, not the current branch/commit).

**Comment by kinow** [comment]
>Btw., is there a possibility to check back, that the latest "version" is actually used (as autosubmit --version will report only the official version, not the current branch/commit).

Good idea! That's done in others tools, and I think we can copy that behaviour. Issue created: https://earth.bsc.es/gitlab/es/autosubmit/-/issues/1324

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Jun 12, 2024, 11:25

Hello @abaer-kit , 

I can't reproduce the exact error, but I found some issues that could affect your case, can you perform git pull and try again?

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Jun 12, 2024, 13:34

mentioned in issue digital-twins/de_340-2/workflow#419

**Comment by kinow** [comment]
In GitLab by @abaer-kit on Jun 12, 2024, 14:13

This did not change anything, unfortunately.

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Jun 12, 2024, 14:15

I just updated that branch and uploaded a new as_conf version (1.0.67), but I'm not sure if it will help you. 

Can you share the contents of conf/metadata/experiment_data.yml? Just in case I'm not replicating the same config this should have the last load info.

Edit: Because I can't have access to that platform, right?

**Comment by kinow** [comment]
In GitLab by @abaer-kit on Jun 12, 2024, 14:24

[experiment_data.yml](https://earth.bsc.es/gitlab/es/autosubmit/uploads/dd6d8e12192e99f3a70d3041d01be94c/experiment_data.yml)

This is for the latest try with new updates and as_conf 1.0.67.

> Because I can't have access to that platform, right?

Not trivially, unfortunately. If this turns out to be a tough thing, I can check, what is possible.

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Jun 12, 2024, 16:36

Hello @abaer-kit ,

Could you try again after a git pull?

**Comment by kinow** [comment]
In GitLab by @abaer-kit on Jun 12, 2024, 17:25

Now it works with this setup both with and without DISABLE_RECOVERY_THREADS.

Also I checked with the slightly complicated workflow (multiple parallel jobs), that I usually use and it works also!

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Jun 13, 2024, 08:24

Perfect! This time, I was almost sure that I got it.

In this version, I added the X11 code, and in the merge on master this line was added: 

```python
chan = self.transport.open_session()
if x11:
...
else:
 chan = self.transport.open_session()
```
I noticed it by going commit by commit. 


So the first `chan = self.transport.open_session()` was opened but never used. This is not an issue in mn5/lumi ( eventually, it would complain about too many fd, but it could take weeks to see it)

Apologies for the issue and the troubleshooting of front and back
